# Типы данных: логические или булевы переменные (boolean)

Следующий тип данных - это логические переменные. Логические переменные могут принимать всего два значения - **истина (`True`)** или **ложь (`False`)**. В Python тип обознчается `bool`.

In [1]:
print(type(True), type(False))

<class 'bool'> <class 'bool'>


Логические переменные чаще всего используется в условных операторах `if-else` и в цикле с остановкой по условию `while`. В части по анализу данных еще увидим одно частое применение - создание булевых масок для фильтрации данных (например, вывести только те данные, где возраст больше 20).

Обратите внимание, что `True` и `False` обязательно пишутся с большой буквы и без кавычек, иначе можно получить ошибку.

In [1]:
print(type('True')) # тип str - строковая переменная
print(true) # ошибка, Python думает, что это назbвание переменной

<class 'str'>


NameError: name 'true' is not defined

Как и в случае чисел и строк, с логическими переменными работает преобразование типов. Превратить что-либо в логическую переменную можно с помощью функции `bool()`.

Преобразование чисел работает следующим образом - 0 превращается в `False`, все остальное в `True`.

In [ ]:
print(bool(0))
print(bool(23))
print(bool(-10))

False
True
True


Пустая строка преобразуется в `False`, все остальные строки в `True`.

In [ ]:
print(bool(''))
print(bool('Hello'))
print(bool(' ')) # даже строка из одного пробела - это True
print(bool('False')) # и даже строка 'False' - это True

False
True
True
True


И при необходимости булеву переменную можно конвертировать в `int`. Тут все без сюрпризов - ноль и единица.

In [ ]:
print(int(True))
print(int(False))

1
0


Ключевое слово `null` обычно используется во многих языках программирования, таких как Java, C++, C# и JavaScript. Это значение, которое присваивается переменной.

Концепция ключевого слова `null` в том, что она дает переменной нейтральное или "нулевое" поведение.

А что же в Python?

Эквивалент `null` в `Python`: `None`
Он был разработан таким образом, по двум причинам:

Многие утверждают, что слово `null` несколько эзотерично. Это не наиболее дружелюбное слово для новичков. Кроме того, `None` относится именно к требуемой функциональности - это ничего, и не имеет поведения.

Присвоить переменной значение `None` очень просто:

In [3]:
my_none_variable = None

In [ ]:
type(None)

NoneType

In [ ]:
bool(None)

False

## Логические выражения

Давайте посмотрим, где используется новый тип данных.

Мы поработаем с логическими выражениями, результат которых - булевы переменные.

Логические выражения выполняют проверку на истинность, то есть выражение равно True, если оно истинно, и False, если ложно.

В логических выражениях используются операторы сравнения:
* ==  (равно)
* !=  (не равно)
* \>  (больше)
* <  (меньше)
* \>=  (больше или равно)
* <=  (меньше или равно)

In [ ]:
print(1 == 1)
print(1 != '1') 
c = 1 > 3
print(c)
print(type(c))
x = 5
print(1 < x <= 5) #можно писать связки цепочкой

True
True
False
<class 'bool'>
True


Логические выражения можно комбинировать с помощью следующих логических операторов:

* логическое И (and) - выражение истинно, только когда обе части истинны, иначе оно ложно
* логическое ИЛИ (or) - выражение ложно, только когда обе части ложны, иначе оно истинно
* логическое отрицание (not) - превращает True в False и наоборот

In [ ]:
print((1 == 1) and ('1' == 1))
print((1 == 1) or ('1' == 1))
print(not(1 == 1))
print((1 == 1) or ('1' == 1) and (2 == 3))

False
True
False
True


# ==> Task 1
## Вася в Италии

Вася уехал учиться по обмену на один семестр в Италию. Единственный магазин в городе открыт с 6 до 8 утра и с 16 до 17 вечера (включительно). Вася не мог попасть в магазин уже несколько дней и страдает от голода. Он может прийти в магазин в X часов. Если магазин открыт в X часов, то выведите `True`, а если закрыт - выведите `False`.

В единственной строке входных данных вводится целое число X, число находится в пределах от 0 до 23

**Формат ввода**  

Целое число X, число находится в пределах от 0 до 23

**Формат вывода**

True или False

#### Примеры

Тест 1  
**Входные данные:**  
16

**Вывод программы:**  
True

# Типы данных: вещественные числа (`float`)

По сути, вещественные числа это десятичные дроби, записанные через точку. Вещественные числа в питоне обозначаются словом float (от "плавающей" точки в них). Также могут быть представлены в виде инженерной записи: 1/10000 = 1e-05

**ВЕЩЕСТВЕННЫЕ ЧИСЛА (FLOAT):** 3.42, 2.323212, 3.0, 1e-05 - число с дробной частью (в том числе целое с дробной частью равной 0)

In [ ]:
4.5 + 5

9.5

Если у нас было действие с участие целого и вещественного числа, то ответ всегда будет в виде вещественного числа (см. выше).

Также давайте вспомним про "обычное" деление, в результате которого получается вещественное число.

In [ ]:
print(11 / 2)
print(type(11 / 2))
print(11 // 2)
print(type(11 // 2))

5.5
<class 'float'>
5
<class 'int'>


С вещественными числами нужно быть осторожными со сравнениями. В связи с особенностями устройства памяти компьютера дробные числа хранятся там весьма хитро и не всегда условные 0.2 то, чем кажутся. Это связано с проблемой точности представления чисел.

Подробнее можно прочитать [здесь](https://pythoner.name/documentation/tutorial/floatingpoint).

In [ ]:
0.2 + 0.1 == 0.3

False

Наверняка, от такого равенства мы ожидали результат True, но нет. Поэтому будьте аккуратны и старайтесь не "завязывать" работу вашей программы на условия, связанные с вычислением вещественных чисел. Давайте посмотрим, как на самом деле выглядят эти числа в памяти компьютера.

In [ ]:
print(0.2 + 0.1)
print(0.3)

0.30000000000000004
0.3


Числа с плавающей точкой представлены в компьютерном железе как дроби с основанием 2 (двоичная система счисления). Например, десятичная дробь

0.125  
имеет значение 1/10 + 2/100 + 5/1000, и таким же образом двоичная дробь

0.001  
имеет значение 0/2 + 0/4 + 1/8. Эти две дроби имеют одинаковые значения, отличаются только тем, что первая записана в дробной нотации по основанию 10, а вторая по основанию 2.

К сожалению, большинство десятичных дробей не могут быть точно представлены в двоичной записи. Следствием этого является то, что в основном десятичные дробные числа вы вводите только приближенными к двоичным, которые и сохраняются в компьютере.

Если вам совсем не обойтись без такого сравнения, то можно сделать так: сравнивать не результат сложения и числа, а разность эти двух чисел с каким-то очень маленьким числом (с таким, размер которого будет точно не критичен для нашего вычисления). Например, порог это числа будет разным для каких-то физических вычислений, где важна высокая точность, и сравнения доходов граждан.

In [ ]:
0.2 + 0.1 - 0.3 < 0.000001

True

Следующая проблема, с которой можно столкнуться - вместо результата вычисления получить ошибку `Numerical result out of range`. Cвязано это с ограничением выделяемой памяти на хранение вещественного числа.

In [ ]:
w = 1.5 ** 100000

OverflowError: ignored

In [ ]:
1.5 ** 1000

1.2338405969061735e+176

А если все получилось, то ответ еще может выглядеть вот так. Такая запись числа называется научной и экономит место - она хранит целую часть числа (мантисса) и степень десятки на которую это число нужно умножить (экспонента). Здесь результатом возведения 1.5 в степень 1000 будет число 1.2338405969061735, умноженное на 10 в степень 176. Понятно, что это число очень большое. Если бы вместо знакак + стоял -, то и степень десятки была бы отрицательная (10 в -176 степени), и такое число было бы очень, очень маленьким.

Как и в случае с целыми числами, вы можете перевести строку в вещественное число, если это возможно. Сделать это можно фукнцией float()

In [ ]:
print(2.5 + float('2.4'))

4.9


## Округление вещественных чисел

У нас часто возникает необходимость превратить вещественное число в целое ("округлить"). В питоне есть несколько способов это сделать, но, к сожалению, ни один из них не работает как наше привычное округление и про это всегда нужно помнить.

Большинство этих функций не реализованы в базовом наборе команд питона и для того, чтобы ими пользоваться, нам придется загрузить дополнительную библиотеку math, которая содержит всякие специальные функции для математических вычислений.

In [ ]:
import math # команда import загружает модуль под названием math

Модуль math устанавливается в рамках дистрибутива Anaconda, который мы использовали, чтобы установить Jupyter Notebook, поэтому его не нужно отдельно скачивать, а можно просто импортировать (загрузить в оперативную память текущей сессии). Иногда нужную библиотеку придется сначала установить на компьютер с помощью команды !pip install -название модуля- и только потом импортировать.

Самый простой способ округлить число - применить к нему функцию int.

In [ ]:
int(2.6)

2

Обратите внимание, что такой метод просто обрубает дробную часть (значения выше 0.5 не округляются в сторону большего числа).

In [ ]:
print(int(2.6))
print(int(-2.6))

2
-2


Округление "в пол" из модуля math округляет до ближайшего меньшего целого числа.

In [ ]:
print(math.floor(2.6)) # чтобы использовать функцю из дополнительного модуля - 
                        # нужно сначала написать название этого модуля и через точку название функции
print(math.floor(-2.6))

2
-3


Округление "в потолок" работает ровно наоброт - округляет до ближайшего большего числа, независимо от значения дробной части.

In [ ]:
print(math.ceil(2.6))
print(math.ceil(-2.6))

3
-2


В самом питоне есть еще функция round(). Вот она работает почти привычно нам, если бы не одно "но"...

In [ ]:
print(round(2.2))
print(round(2.7))
print(round(2.5)) # внимание на эту строку
print(round(3.5))

2
3
2
4


Неожиданно? Тут дело в том, что в питоне реализованы не совсем привычные нам правила округления чисел с вещественной частью 0.5 - такое число округляется до ближайшего четного числа: 2 для 2.5 и 4 для 3.5.

## Замечание по импорту функций
Иногда нам не нужна вся библиотека, а только одна функция из-за нее. Скажите, странно же хранить в опреативной памяти всю "Войну и мир", если нас интересует только пятое предложение на восьмой странице.
Для этого можно воспользоваться импортом функции из библиотеки и тогда не нужно будет писать ее название через точку. Подводный камень здесь только тот, что если среди базовых команд питона есть функция с таким же именем, то она перезапишется и придется перезапускать свой блокнот, чтобы вернуть все как было.

In [ ]:
from math import ceil
ceil(2.6) # теперь работает без math.

3

In [ ]:
1+3j

(1+3j)

# ==> Task 2
## Дробная часть

Дано вещественное число. Выведите его дробную часть.

**Формат ввода**  

Вещественное число

**Формат вывода**

Вещественное число (ответ на задачу)

#### Примеры

Тест 1  
**Входные данные:**  
4.0

**Вывод программы:**  
0.0

## Строки. Ввод и форматирование.

**ТЕКСТ (СТРОКИ) (STR, STRING):** любой текст внутри одинарных или двойных кавычек. 
Важно: целое число в кавычках - это тоже строка, не целое число.

In [1]:
x = 'text'
print(type(x))
print(type('Hello, world!'))
print(type('2'))

<class 'str'>
<class 'str'>
<class 'str'>


Если попробовать сложить число-строку и число-число, то получим ошибку.

In [2]:
print('2' + 3)

TypeError: must be str, not int

Еще раз обратим внимание на текст ошибки. *TypeError: must be str, not int* - TypeError означает, что данная команда не может быть исполнена из-за типа какой-то из переменных. В данном случаем видим, что что-то должно быть "str", когда оно типа "int". Давайте попробуем сделать 3 тоже строкой. Кстати, неважно какие вы используете кавычки (только если это не текст с кавычками внутри), главное, чтобы открывающая и закрывающая кавычки были одинаковые.

In [3]:
print('2' + "3")

23


Такая операция называется контакенация (слиянием) строк. 

Можно и умножить строку на число можно. Такая операция повторит нам строку заданное количество раз.

In [4]:
'2' * 3

'222'

Операции со строками тоже работают, если строки лежат внутри переменных.

In [5]:
word1 = 'John'
word2 = ' Brown'
print(word1 + word2)
word3 = word1 + word2 # можем результат контакенации положить в новую переменную
print(word3)

John Brown
John Brown


В прошлом блокноте мы преобразовывали строки в числа с помощью функции int(). 

In [6]:
print(2 + int('2512')) # нет ошибки!

2514


С помощью функции str() мы можем превращать другие типы данных в строки.

In [7]:
print('abs' + str(123) + str(True))

abs123True


Обратите внимание: эти функции (str, int, float и другие) не меняют тип самих данных или переменных, в которых они хранятся. Если мы не перезапишем значение, то строка станет числом только в конкретной строчке команды, а ее тип не изменится.

In [8]:
a = '2342123'
print(type(a))
print(2 + int(a))
print(type(a))

<class 'str'>
2342125
<class 'str'>


In [9]:
a = int(a) # перезаписываем значение переменной
print(type(a)) # теперь изменился и тип

<class 'int'>


# Ввод данных. input()

Познакомимся с функцией input() - ввод. Это функция позволяет нам просить
пользователя что-нибудь ввести или принять на ввод какие-то данные.

In [10]:
x = input('Как вас зовут?')
print(x)

Как вас зовут?Semen
Semen


'Как вас зовут?' - аргумент функции input(), который будет отображаться в строке ввода. Он необязателен, но он может быть полезен для вас (особенно, если в задаче требуется ввести больше одной строки данных и нужно не запутаться).

Обратите внимание, что функция  input() все сохраняет в строковом формате данных. Но мы можем сделать из
числа-строки число-число с помощью функции int(). Введите "1" в обоих случаях и сравните результат.

In [11]:
print(type(input()))
print(type(int(input())))

Sem
<class 'str'>
123
<class 'int'>


Нужно следить за тем, какие данные вы считываете и менять их при необходимости. Сравните два примере ниже.

In [12]:
x = input('Input x value: ')
y = input('Input y value: ')
print(x + y) # произошла склейка строк, по умолчанию input() сохраняет значение типа str

Input x value: 2
Input y value: 3
23


In [13]:
x = int(input('Input x value: '))
y = int(input('Input y value: '))
print(x + y) # вот теперь произошло сложение чисел

Input x value: 2
Input y value: 3
5


# ==> Task 3
## Pig Latin 

Pig Latin - вымышленный язык, который в конец каждого слова подставляет 'yay' (на самом деле оригинальные условия задачи несколько хитрее, но мы будем к ней возвращаться по ходу курса.

**Формат ввода**  
Пользователь вводит слова.

**Формат вывода**

Слово на Pig Latin.


#### Примеры
Тест 1  
**Ввод:**  
apple 

**Вывод:**  
appleyay

In [14]:
# Your code

# ==> Task 4


## x потвторенное x раз в квадрате

1) Вводится целое число X. 

2) $ (X * X)^2 = result$ 

3) Напечатайте получившееся.


**Формат ввода**  
Целое число

**Формат вывод**  
Ответ на задачу

#### Примеры
Тест 1  
**Ввод:**  
2  

**Вывод:**  
484

In [15]:
# Your code

# ==> Task 5 
## Хитрости умножения  

Для умножения двузначного числа на 11 есть интересная хитрость: результат произведения можно получить если между первой и второй цифрой исходного числа вписать сумму этих цифр. Например, 15 * 11 = 1 1+5 5 = 165 или 34 * 11 = 3 3+4 4 = 374. Реализуйте программу, которая умножала бы двузначные числа на 11, используя эту хитрость. Пользоваться оператором умножения нельзя.  

__Формат ввода:__  
Вводится двузначное число.  
__Формат вывода:__  
Программа должна вывести ответ на задачу.  

**Тест 1**  
__Пример ввода:__  
15   
__Вывод:__  
165  

**Тест 2**  
__Пример ввода:__   
66      
__Вывод:__   
726   

In [19]:
# Your code

## Форматирование строк (string formatting)
А теперь посмотрим на то, как подставлять значения в уже имеющийся текстовый шаблон, то есть форматировать строки. Чтобы понять, о чём идет речь, можно представить, что у нас есть электронная анкета, которую заполняет пользователь, и мы должны написать программу, которая выводит на экран введенные данные, чтобы пользователь мог их проверить.

Пусть для начала пользователь вводит свое имя и возраст.

In [21]:
name = input("Введите Ваше имя: ")
age = int(input("Введите Ваш возраст: ")) # возраст будет целочисленным

Введите Ваше имя: Ян
Введите Ваш возраст: 28


Теперь выведем на экран сообщение вида

`Ваше имя: `имя`. Ваш возраст: `возраст`.`

Но прежде, чем это сделать, поймем, какого типа будут значения, которые мы будем подставлять в шаблон. Имя (переменная name) – это строка (string), а возраст (переменная age) – это целое число (integer).

In [26]:
str(28)

'28'

## Option #1

In [27]:
result = "Ваше имя: %s. Ваш возраст: %i." % (name, age)
print(result)

Ваше имя: Ян. Ваш возраст: 28.


Что за таинственные %s и %i? Все просто: оператор % в строке указывает место, на которое будет подставляться значение, а буква сразу после процента – сокращённое название типа данных (s – от string и i – от integer). Осталось только сообщить Python, что именно нужно подставлять – после кавычек поставить % и в скобках перечислить названия переменных, значения которых мы будем подставлять.

Примечание: не теряйте часть с переменными после самой строки. Иначе получится нечто странное:

In [ ]:
print("Ваше имя: %s. Ваш возраст: %i.")

Ваше имя: %s. Ваш возраст: %i.


Важно помнить, что если мы забудем указать какую-то из переменных, мы получим ошибку (точнее, исключение): Python не будет знать, откуда брать нужные значения.

In [28]:
print("Ваше имя: %s. Ваш возраст: %i." % (name))

TypeError: ignored

Кроме того, создавая такие текстовые шаблоны, нужно обращать внимание на типы переменных, значения которых мы подставляем.

In [30]:
print("Ваше имя: %s. Ваш возраст: %s." % (name, age)) # так сработает

Ваше имя: Ян. Ваш возраст: 28.


In [ ]:
print("Ваше имя: %i. Ваш возраст: %s." % (name, age)) # а так нет

TypeError: %i format: a number is required, not str

В первом случае код сработал: Python не очень строго относится к типам данных, и поэтому он легко может превратить целочисленный возраст в строку (два %s вместо %s и %i не является помехой). Во втором случае все иначе. Превратить строку, которая состоит из букв (name) в целое число никак не получится, поэтому Python справедливо ругается.

А что будет, если мы будем подставлять не целое число, а дробное, с плавающей точкой? Попробуем!

In [42]:
height = float(input("Введите Ваш рост (в метрах): "))
height

Введите Ваш рост (в метрах): 1.82


1.82

In [38]:
print("Ваш рост: %f м." % height) # f - от float

Ваш рост: 2.500000 м.


По умолчанию при подстановке значений типа float Python выводит число с шестью знаками после запятой. Но это можно исправить. Перед f нужно поставить точку и указать число знаков после запятой, которое мы хотим:

In [33]:
print("Ваш рост: %.2f м." % height) # например, два

Ваш рост: 1.82 м.


In [40]:
print("Ваш рост: %.0f м. " % height) # или один

Ваш рост: 2 м. 


## Option #2

В случае, если указанное число знаков после запятой меньше, чем есть на самом деле (как в ячейке выше), происходит обычное арифметическое округление.

Рассмотренный выше способ форматирования строк – не единственный. Он довольно стандартный, но при этом немного устаревший. В Python 3 есть другой способ ‒ форматирование с помощью метода .format(). Кроме того, в Python 3.6 и более поздних версиях появился еще более продвинутый способ форматирования строк ‒ f-strings (formatted string literals).

F-strings очень удобны и просты в использовании: вместо % и сокращённого названия типа в фигурных скобках внутри текстового шаблона нужно указать название переменной, из которой должно подставляться значение, а перед всей строкой добавить f, чтобы Python знал, что нам нужна именно f-string.

In [ ]:
print(f"Ваше имя: {name}. Ваш возраст: {age}. Рост: {height:.2f}")

Ваше имя: Олег. Ваш возраст: 32. Рост: 1.78


Альтерантивой такого кода будет следующий синтаксис. Здесь переменные вставятся в порядке "упоминания" в строке.

## Option #3

In [41]:
print("Ваше имя: {}. Ваш возраст: {}".format(name, age))

Ваше имя: Ян. Ваш возраст: 28


Форматирование дробей добавляем так.

In [ ]:
print("Ваше имя: {}. Ваш возраст: {}. Рост: {:.2f}".format(name, age, height))

Ваше имя: Олег. Ваш возраст: 32. Рост: 1.78


Если указали формат переменной (.2f, например, ожидает float), то следим за порядком.

In [44]:
print("Ваше имя: {}. Ваш возраст: {}. Рост: {:.2f}".format(age, height, name)) 

ValueError: ignored

Порядок заполнения можно указывать с помощью нумерации. Одну и тоже переменную можно использовать несколько раз.

In [45]:
print("Ваше имя: {2}. Ваш возраст: {0}. Рост: {1:.2f}. Все верно, {2}?".format(age, height, name)) 

Ваше имя: Ян. Ваш возраст: 28. Рост: 1.82. Все верно, Ян?


Также при форматировании строк можно использовать результаты вычислений и результаты работы методах (о  них немного позже). name.upper() в примере делает все символы строки name заглавными.

In [46]:
print(f"Ваше имя: {name.upper()}. Ваш возраст: {2021-1993}. Рост в футах: {height * 100 / 30.48 :.2f}")

Ваше имя: ЯН. Ваш возраст: 28. Рост в футах: 5.97


# ==> Task 6
## Pi

Из модуля math импортируйте переменную pi. С помощью %f (первое предложение) и format (второе предложение) форматирования выведите строки из примера. Везде, где встречается число pi - это должна быть переменная pi, округленная до определенного знака после точки. 

 
__Формат вывода:__  
`Значение 22/7 (3.14) является приближением числа pi (3.1416)`  
`Значение 22/7 3.142 является приближением числа pi 3.141592653589793`



In [20]:
import math

# Your code

# Функция print(): аргументы и параметры

И еще немного поговорим про форматирование вывода.На самом деле функция `print()` немного сложнее, чем то, что мы уже видели. Например, мы можем печатать одной функцией больше чем один аргумент и использовать разделители вместо пробелов. 

In [47]:
print('2 + 3 =', 2 + 3)
z = 5
print('2 + 3 =', z)

2 + 3 = 5
2 + 3 = 5


У функции `print` есть параметры. Параметры - это такие свойства функций, которые задают значение невидимых нам аргументов внутри функции. Так существует параметр sep (separator), благодаря которому мы можем менять тип разделителя между аргументами `print`. В качестве разделителя может выступать любая строка.  
Сравните:

In [21]:
print('1', '2', '3')
print()
print('1', '2', '3', sep='.')
print()
print('1', '2', '3', sep='')
print()
print('1', '2', '3', sep='\n')

1 2 3

1.2.3

123

1
2
3


Параметр `end` задает то, что будет выведено в конце исполнения функции `print`. По умолчанию там стоит 
невидимый символ `\n`, который осуществляет переход на новую строку. Мы можем заменить его на любую строку. 
Если мы хотим сохранить переход на новую строку - то обязательно прописываем наш невидимый символ внутри
выражения.

In [51]:
print('1', '2', '3', sep='.', end='!')
print('2')  # строки слились

print('1', '2', '3', sep='.', end='!\n')
print('2')  # вывод на новой строке


1.2.3!2
1.2.3!
2


# Home tasks

# Task #1

Create a program that asks the user to enter their name and their age. Print out a message addressed to them that tells them the year that they will turn 100 years old.


__Output:__

`
Your name: Sem
How old are you: 25
Sem will be 100 years old in the year 2096
`

In [27]:
# Your code